test/system/include/metashell_system_test/pattern.hpp

```diff
@@ -0,0 +1,91 @@
+#ifndef METASHELL_SYSTEM_TEST_PATTERN_HPP
+#define METASHELL_SYSTEM_TEST_PATTERN_HPP
+
+// Metashell - Interactive C++ template metaprogramming shell
+// Copyright (C) 2015, Abel Sinkovics (abel@sinkovics.hu)
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+#include <metashell_system_test/placeholder.hpp>
+
+#include <boost/variant.hpp>
+#include <boost/optional.hpp>
+
+#include <string>
+
+namespace metashell_system_test
+{
+  template <class T>
+  class pattern
```

Comment:
I've added comments explaining it in 4f9639635af54179dc6a221ffb62a0232bf4173d. Let me know if this is still not clear.